src/ffi/ios.rs

```diff
@@ -2,41 +2,110 @@ use core_foundation::{
     base::TCFType,
     string::{CFString, CFStringRef},
 };
-use http::Request;
+use http::{Request, HeaderValue, HeaderMap, StatusCode};
 use libc::c_char;
+use tracing::{info, debug};
+use url::Url;
+use anyhow::Result;
+use serde::{Serialize, Deserialize};
+use http::header::HeaderName;
+use std::collections::HashMap;
 
 use crate::{client::Client, DirectoryCache};
+use std::{fs, io};
 
-fn setup_logger() {
-    let subscriber = tracing_fmt::FmtSubscriber::builder()
-        .with_max_level(tracing::Level::DEBUG)
-        .with_writer(std::io::stderr)
-        .finish();
-    tracing::subscriber::set_global_default(subscriber).expect("to be the only logger");
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ArtiRequest {
+    method: String,
+    url: String,
+    headers: HashMap<String, Vec<String>>,
+    body: Vec<u8>,
+    dict_dir: String,
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn call_tls_get(domain_cc: *const c_char) -> CFStringRef {
-    let domain = cstring_to_str(&domain_cc);
+pub unsafe extern "C" fn call_arti(
+    request_json: *const c_char) -> CFStringRef {
+    setup_logger();
+
+    let ret = match _call_arti(cstring_to_str(&request_json)) {
+        Ok(ret) => ret,
+        Err(e) => ReturnStruct::err(e.to_string()),
+    };
+    return to_cf_str(ret.to_json().to_string());
+}
 
-    let client = Client::new(DirectoryCache {
-        tmp_dir: None,
+pub fn _call_arti(request_json: &str) -> Result<ReturnStruct> {
+    let request: ArtiRequest = serde_json::from_str(request_json)?;
+    info!("Request is: {:?}", request);
+
+    let host = Url::parse(&request.url)?.host_str().unwrap().to_string();
+    let mut req = match request.method.as_str() {
+        "GET" => Request::get(request.url),
+        "POST" => Request::post(request.url),
+        "PUT" => Request::put(request.url),
+        "DELETE" => Request::delete(request.url),
+        // TODO: add other verbs of REST
+        // TODO: correctly return error here
+        _ => Request::get(request.url),
+    }
+        .header("Host", host)
+        .version(http::Version::HTTP_10)
+        .body(vec![])?;
```

Comment:
I found that you can use `http::method::Method::from_bytes` to automatically parse the method.
I've also added a quick example on how to better handle error: don't use `unwrap` or a `?` without `context`. `anyhow::Context` adds `context()` on both `Option` and `Result` so you can easily add some info before returning.

```suggestion
    let host = Url::parse(&request.url)
        .context("parse url")?
        .host_str()
        .context("no host in request")?;

    let mut req = Request::builder()
        .method(Method::from_bytes(request.method.as_bytes()).context("invalid method")?)
        .header("Host", host)
        .version(http::Version::HTTP_10)
        .body(vec![])
        .context("invalid request")?;
```